### PR TITLE
[VSPC-170] Implemented the changes so that block order in slides rema…

### PR DESCRIPTION
**Pull request by AvirupBiswas**
…in consistent even after deletion

# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
Slides: block order isn't handled correctly

After fix:
Slide block's order remains consistent after deletion.
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ContentBlockRepository.java [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 
 @Repository
 @JaversSpringDataAuditable
 public interface ContentBlockRepository extends PagingAndSortingRepository<ContentBlock, String> {
 
+    @Query("SELECT max(contentOrder) FROM ContentBlock d WHERE d.slide.id = ?1")
+    public Integer findMaxContentOrder(String slideId);
+    
+    @Modifying
+    @Transactional
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ContentBlockRepository.java:19]: transactional should be on the manger method not the repository

File: vspace/src/main/java/edu/asu/diging/vspace/core/data/ContentBlockRepository.java [hunk 2]
```diff
@@ -1,13 +1,23 @@
...
+    @Query(value="UPDATE ContentBlock c SET c.contentOrder = c.contentOrder -1 WHERE c.contentOrder > (SELECT d.contentOrder FROM ContentBlock d WHERE d.id=?1)")
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/data/ContentBlockRepository.java:20]: this is business logic that should go in the manager

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddChoiceBlockController.java [hunk 3]
```diff
@@ -25,21 +26,33 @@
...
-        Map<String,Object> responseData = new HashMap<String,Object>();
-        responseData.put("choiceBlock",choiceBlock);
-        if(!choiceBlock.isShowsAll()) {
-            List<ISequence> selectedSequences = choiceBlock.getChoices().stream().map(choice->choice.getSequence()).collect(Collectors.toList());
-            responseData.put("selectedSequences",selectedSequences);
+        Integer contentOrder = contentBlockRepository.findMaxContentOrder(slideId);
+        if (contentOrder == null) {
+            contentOrder = -1;
+        }
+        contentOrder=contentOrder+1;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddChoiceBlockController.java:41]: merge lines 38 - 41 with ternary operator

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddChoiceBlockController.java [hunk 4]
```diff
@@ -25,21 +26,33 @@
     @Autowired
     private IContentBlockManager contentBlockManager;
 
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddChoiceBlockController.java:30]: controllers should not directly call repositories (that's the job of the manager/service classes)

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddImageBlockController.java [hunk 5]
```diff
@@ -28,11 +29,13 @@
     @Autowired
     private IContentBlockManager contentBlockManager;
 
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddImageBlockController.java:33]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddImageBlockController.java [hunk 6]
```diff
@@ -41,6 +44,11 @@
             filename = file.getOriginalFilename();
         }
         String imageId;
+        Integer contentOrder = contentBlockRepository.findMaxContentOrder(slideId);
+        if (contentOrder == null) {
+            contentOrder = -1;
+        }
+        contentOrder=contentOrder+1;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddImageBlockController.java:51]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddTextBlockController.java [hunk 7]
```diff
@@ -20,11 +21,18 @@
     @Autowired
     private IContentBlockManager contentBlockManager;
 
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddTextBlockController.java:25]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddTextBlockController.java [hunk 8]
```diff
@@ -20,11 +21,18 @@
...
     public ResponseEntity<String> addTextBlock(@PathVariable("id") String slideId,
-            @PathVariable("moduleId") String moduleId, @RequestParam("content") String content,
-            @RequestParam("contentOrder") Integer contentOrder) throws IOException {
+            @PathVariable("moduleId") String moduleId, @RequestParam("content") String content) throws IOException {
 
+        Integer contentOrder = contentBlockRepository.findMaxContentOrder(slideId);
+        if (contentOrder == null) {
+            contentOrder = -1;
+        }
+        contentOrder=contentOrder+1;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddTextBlockController.java:35]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteChoiceBlockController.java [hunk 9]
```diff
@@ -23,9 +24,13 @@
     @Autowired
     private IContentBlockManager contentBlockManager;
 
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteChoiceBlockController.java:28]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteImageBlockController.java [hunk 10]
```diff
@@ -22,10 +23,14 @@
 
     @Autowired
     private IContentBlockManager contentBlockManager;
+    
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteImageBlockController.java:28]: see above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteTextBlockController.java [hunk 11]
```diff
@@ -22,11 +23,16 @@
 
     @Autowired
     private IContentBlockManager contentBlockManager;
-
+    
+    @Autowired
+    private ContentBlockRepository contentBlockRepository;
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteTextBlockController.java:28]: see above

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java [hunk 12]
```diff
@@ -253,17 +286,42 @@ public IChoiceBlock getChoiceBlock(String choiceBlockId) {
...
+     * specified contentOrder
+     */
+    private void updateContentOrder(Integer contentOrder) {
+        if (contentOrder != null) {
+            List<ContentBlock> contentBlockList = contentBlockRepository.findByContentOrderGreaterThan(contentOrder);
+            for (ContentBlock eachContentBlock : contentBlockList) {
+                eachContentBlock.setContentOrder(eachContentBlock.getContentOrder() - 1);
+            }
+            contentBlockRepository.saveAll(contentBlockList);
+        }
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java:325]: doesn't thism method update the content order for all content blocks no matter to which slide they belong? Not just the content blocks of a specific slide? that seems wrong.

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java (hunk 12 above)
Comment (AvirupBiswas) [reply vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java:327]: Yes. Slideid got missed.Updated the code.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java [hunk 13]
```diff
@@ -253,17 +287,43 @@ public IChoiceBlock getChoiceBlock(String choiceBlockId) {
...
+     */
+    @Override
+    public Integer findMaxContentOrder(String slideId) {
+        return contentBlockRepository.findMaxContentOrder(slideId);
+    }
+
+    /**
+     * decreasing content order by 1 of the slide's block which are after the
+     * specified contentOrder
+     */
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java:318]: should be proper javadoc (same for above method)

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java [hunk 14]
```diff
@@ -253,17 +287,43 @@ public IChoiceBlock getChoiceBlock(String choiceBlockId) {
...
+    private void updateContentOrder(String slideId,Integer contentOrder) {
+        
+        if (slideId!=null && contentOrder != null) {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManager.java:321]: can this ever happen? it seems overkill for this method.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
test cases are failing

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java [hunk 15]
```diff
@@ -22,51 +34,62 @@
...
-
+        contentBlock = new ContentBlock();
+        contentBlock.setContentOrder(1);
     }
 
     @Test
     public void test_deleteTextBlockById_success() throws BlockDoesNotExistException {
         String textBlockId = "2";
-        managerToTest.deleteTextBlockById(textBlockId);
+        Optional<ContentBlock> ContentBlockOptional = Optional.of(contentBlock);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java:49]: variable names are lowercased in java!

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java [hunk 16]
```diff
@@ -22,51 +34,62 @@
...
+        when(contentBlockRepository.findById("2")).thenReturn(ContentBlockOptional);
+        managerToTest.deleteTextBlockById(textBlockId,null);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java:51]: the test should get real data (not null) and also test the additional functionality (the logic you've added)

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java [hunk 17]
```diff
@@ -22,51 +34,62 @@
...
         Mockito.doThrow(EmptyResultDataAccessException.class).when(imageBlockRepo).deleteById(imageBlockId);
-        managerToTest.deleteImageBlockById(imageBlockId);
+        managerToTest.deleteImageBlockById(imageBlockId,null);
     }
 
     @Test
     public void test_deleteImagetBlockById_whenIdIsNull() throws BlockDoesNotExistException {
         String imageBlockId = null;
-        managerToTest.deleteImageBlockById(null);
+        managerToTest.deleteImageBlockById(null,null);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/ContentBlockManagerTest.java:92]: see above

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Pull request merged**